src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,221 @@
+                            if (asn.as_value() == "AS32934") {
+                                (*fb_service_ips)[service].push_back(answer.ipv4);
+                            } else {
+                                (*entry)["facebook_dns_blocking"] = true;
+                            }
+                        }
+                    }
+                }
+                // if any consistent IPs, consistent = true
+                (*entry)["facebook_" + service + "_dns_consistent"] =
+                    !(*fb_service_ips)[service].empty();
+            }
+            *names_tested += 1;
+            if (names_count == *names_tested) {
+                cb(NoError(), entry, *fb_service_ips, options, reactor, logger);
+                return;
+            }
+        };
+    };
+
+    for (auto const& service_and_hostname : fb_service_hostnames) {
+        std::string service = service_and_hostname.first;
+        std::string hostname = service_and_hostname.second;
+        templates::dns_query(entry, "A", "IN", hostname, "",
+                             dns_cb(service, hostname), options, reactor,
+                             logger);
+    }
+    return;
+}
+
+static void tcp_many(Error error,
+                     Var<Entry> entry,
+                     std::map<std::string, std::vector<std::string>> fb_service_ips,
+                     Settings options,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Var<Entry>> cb
+                    ) {
+    logger->info("starting tcp_many");
+    if (error) {
+        cb(entry);
+        return;
+    }
+
+    // if we find any blocked TCP later, switch this to true
+    (*entry)["facebook_dns_blocking"] = false;
+
+    int ips_count = 0;
+    for (auto const& service_and_ips : fb_service_ips) {
+        // skipping stun for now
+        if (service_and_ips.first == "stun") {
+            continue;
+        }
+        ips_count += service_and_ips.second.size();
+    }
+    if (ips_count == 0) {
+        cb(entry);
+        return;
+    }
+    Var<int> ips_tested(new int(0));
+
+    auto tcp_cb = [=](std::string service, std::string ip, int port) {
+        return [=](Error err, Var<net::Transport> txp) {
+            bool close_txp = true; // if connected, we must disconnect
+            if (!!err) {
+                logger->info("tcp failure to %s at %s:%d", service.c_str(),
+                    ip.c_str(), port);
+                (*entry)["facebook_" + service + "_blocking"] = true;
+                (*entry)["facebook_dns_blocking"] = true;
+                close_txp = false;
+            } else {
+                logger->info("tcp success to %s at %s:%d", service.c_str(),
+                    ip.c_str(), port);
+                (*entry)["facebook_" + service + "_blocking"] = false;
+            }
+            *ips_tested += 1;
+            if (ips_count == *ips_tested) {
```

Comment:
Also here, can we `assert` that `ips_count <= *ips_tested` and make sure `ips_count` is `size_t`?